### PR TITLE
Customizable consensus engine

### PR DESCRIPTION
**Pull request by gorbak25**
This feature is in very early stages and development is rather slow, I'm only opening a PR so It's clear for my teammates on what I'm working on right now.

### CODE REVIEW

**Review by erlmachinedev — COMMENTED** [review]
Great job, I think that makes a large impact on application design and some documentation could be suitable to support new developers. (if agree I can help with UML schemas later)

**Review by erlmachinedev — DISMISSED** [review]

**Review by velzevur — APPROVED** [review]
Looks good. I've also ran a quick test if this PR syncs with mainnet and synced all the way to generation `42820` without an issue, then I stopped it.

File: apps/aecore/src/aec_conductor.erl [hunk 1]
```diff
@@ -1132,11 +1204,13 @@ handle_add_block(Block, Hash, Prev, #state{top_block_hash = TopBlockHash} = Stat
...
-    {ok, State};
+    State1 = maybe_consensus_change(State, Block),
+    {ok, State1};
 handle_successfully_added_block(Block, Hash, true, PrevKeyHeader, Events, State, Origin) ->
     maybe_publish_tx_events(Events, Hash, Origin),
     maybe_publish_block(Origin, Block),
-    case preempt_on_new_top(State, Block, Hash, Origin) of
+    State1 = maybe_consensus_change(State, Block),
+    case preempt_on_new_top(State1, Block, Hash, Origin) of
         {micro_changed, State2 = #state{ consensus = Cons }} ->
```
Comment (uwiger) [inline apps/aecore/src/aec_conductor.erl:1214]: Is this correct (creating an entirely new state rather than building on `State1`)?

**Review by uwiger — DISMISSED** [review]

File: apps/aecore/src/aec_conductor.erl (hunk 1 above)
Comment (gorbak25) [reply apps/aecore/src/aec_conductor.erl:1214]: This is a pattern not a expression

**Review by velzevur — APPROVED** [review]

**Review by radrow — APPROVED** [review]

**Comment by radrow** [comment]
Congrats on this!

**Pull request merged**